src/Controls/src/Core/Compatibility/Handlers/Shell/iOS/ShellSectionRenderer.cs

```diff
@@ -109,7 +107,37 @@ public bool ShouldPopItem(UINavigationBar _, UINavigationItem __)
 		[Export("navigationBar:didPopItem:")]
 		[Internals.Preserve(Conditional = true)]
 		bool DidPopItem(UINavigationBar _, UINavigationItem __)
-			=> _popRequested || SendPop();
+		{
+			// Check for null references
+			if (_shellSection?.Stack == null || NavigationBar?.Items == null)
+				return true;
+
+			// Check if stacks are in sync
+			if (_shellSection.Stack.Count == NavigationBar.Items.Length)
+				return true;
+
+			var pages = _shellSection.Stack.ToList();
+
+			// Ensure we have enough pages and navigation items
+			if (pages.Count == 0 || NavigationBar.Items.Length == 0)
+				return true;
+
+			// Bounds check: ensure we have a valid index for pages array
+			int targetIndex = NavigationBar.Items.Length - 1;
+			if (targetIndex < 0 || targetIndex >= pages.Count)
+				return true;
+
+			_shellSection.SyncStackDownTo(pages[targetIndex]);
```

Comment:
Can directly use:
` _shellSection.SyncStackDownTo(_shellSection.Stack[targetIndex]);`
 And avoid the memory allocation from `ToList()`?

Reply:
From line 119

Reply:
The contents of `_shellSection.Stack` is going to change throughout the method so we need to keep a copy of it for the for loop after line 130.

So we might as well just use pages here because we need to allocate the shallow copy of the list anyway.

```C#
for (int i = pages.Count - 1; i >= NavigationBar.Items.Length; i--)
			{
				var page = pages[i];
				if (page != null)
					DisposePage(page);
			}
```